README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# impac

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/rdinnager/impac/workflows/R-CMD-check/badge.svg)](https://github.com/rdinnager/impac/actions)
<!-- badges: end -->

The goal of `{impac}` is to create packed image mosaics. The main
function `impac`, takes a set of images, or a function that generates
images and packs them into a larger image as tightly as possible,
scaling as necessary, using a greedy algorithm (so don’t expect it to be
fast\!). It is inspired by [this python
script](https://github.com/qnzhou/Mosaic%5D). The main upgrade in this
package is the ability to feed the algorithm a generator function, which
generates an images, as opposed to just a list of pre-existing images
(though it can do this too).

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("rdinnager/impac")
```

## Example

This document and hence the images below are regenerated once a day
automatically. No two will ever be alike.

First we load the packages we need for these examples:

``` r
library(impac)
library(Rvcg)
library(rgl)
library(rphylopic)
```

Next we create an R function to generate an image. In this case, we use
the package `rgl` to plot a simple 3d shape, chosen randomly from a set
of possibilities:

``` r

generate_platonic <- function(i, swidth = 200, sheight = 200, cols = rainbow(100)) {
  
  shape <- sample(c("sphere",
                    "spherical_cap",
                    "tetrahedron",
                    "dodecahedron",
                    "octahedron",
                    "icosahedron",
                    "hexahedron",
                    "cube",
                    "cone"),
                  1)
  
  mesh <- switch (shape,
    sphere = Rvcg::vcgSphere(),
    spherical_cap = Rvcg::vcgSphericalCap(),
    tetrahedron = Rvcg::vcgTetrahedron(),
    dodecahedron = Rvcg::vcgDodecahedron(),
    octahedron = Rvcg::vcgOctahedron(),
    icosahedron = Rvcg::vcgIcosahedron(),
    hexahedron = Rvcg::vcgHexahedron(),
    cube = Rvcg::vcgBox(),
    cone = Rvcg::vcgCone(2, 0, 6)
  )
  
  scales <- c(1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4)
  mesh <- rgl::scale3d(mesh, 
                       sample(scales, 1),
                       sample(scales, 1),
                       sample(scales, 1))
  
  mesh <- rgl::rotate3d(mesh, runif(1, 0, 2 * pi), 0, 0, 1)
  mesh <- rgl::rotate3d(mesh, runif(1, 0, 2 * pi), 0, 1, 0)
  mesh <- rgl::rotate3d(mesh, runif(1, 0, 2 * pi), 1, 0, 0)
  
  rgl::shade3d(mesh, col = sample(cols, 1),
               specular = "grey")
  
  png_file <- tempfile(fileext = ".png")
  rgl::snapshot3d(filename = png_file, width = swidth, height = sheight,
                  webshot = FALSE)
  rgl::close3d()
  
  im2 <- imager::load.image(png_file)
  im <- imager::imfill(swidth, sheight, val = c(0, 0, 0, 1))
  im[ , , , 1:3] <- im2 
  im[imager::R(im) == 1 & imager::G(im) == 1 & imager::B(im) == 1] <- 0
  
  im  
 
}
```

Now we feed our function to the `impac()` function, which packs the
generated images onto a canvas:

``` r
shapes <- impac(generate_platonic, progress = FALSE, show_every = 0, bg = "white")
imager::save.image(shapes$image, "man/figures/R_gems.png")
```

![Pretty R gems - Packed images of 3d shapes drawn with
{rgl}](man/figures/R_gems.png)

Now let’s pack some Phylopic images\! These are silhouettes of organisms
from the [Phylopic](http://phylopic.org/) project. We will use the
`rphylopic` package to grab a random Phylopic image for packing:

``` r
all_images <- rphylopic::image_list(1, 10000)
all_images <- unlist(all_images)
get_phylopic <- function(i, max_size = 400, isize = 1024) {
  fail <- TRUE
  while(fail) {
    uuid <- sample(all_images, 1)
    pp <- try(rphylopic::image_data(uuid, isize), silent = TRUE)
    if(!inherits(pp, "try-error")) {
      fail <- FALSE
    }
  }
  rot <- aperm(pp$uid, c(2, 1, 3))
  dims <- dim(rot)
  im <- imager::as.cimg(as.vector(rot), dim = c(dims[1], dims[2], 1, dims[3]))
  max_dim <- which.max(dims[1:2])
  other_dim <- (max_size / dims[max_dim]) * dims[1:2][-max_dim]
  new_size <- c(0, 0)
  new_size[max_dim] <- max_size
  new_size[-max_dim] <- other_dim
  im <- imager::resize(im, new_size[1], new_size[2], interpolation_type = 6)
  im <- imager::imchange(im, ~ . < 0, ~ 0)
  im <- imager::imchange(im, ~ . > 1, ~ 1)
  ## this adds custom metadata
  list(im, uuid = uuid)
}
```

Now we run `impac` on our phylopic generating function:

``` r
phylopics <- impac(get_phylopic, progress = FALSE, show_every = 0, bg = "white", min_scale = 0.01)
imager::save.image(phylopics$image, "man/figures/phylopic_a_pack.png")
```

![Packed images of organism silhouettes from
Phylopic](man/figures/phylopic_a_pack.png)

Now we extract the artists who made the above images using the uid of
image.

``` r
image_dat <- lapply(phylopics$meta$uuid, 
                    function(x) {Sys.sleep(2); rphylopic::image_get(x, options = c("credit"))$credit})
```

## Artists whose work is showcased:

T. Michael Keesey, Arthur S. Brum, Julio Garza, Birgit Lang, Elisabeth
Östman, Mo Hassan, Scott Hartman, Zimices, Matt Crook, Steven Traver,
Margot Michaud, Matthew E. Clapham, Sergio A. Muñoz-Gómez, Nobu Tamura,
vectorized by Zimices, Didier Descouens (vectorized by T. Michael
Keesey), Chase Brownstein, Tracy A. Heath, Christoph Schomburg, Mali’o
Kodis, drawing by Manvir Singh, Walter Vladimir, Ray Simpson (vectorized
by T. Michael Keesey), Chris huh, Gabriela Palomo-Munoz, mystica,
Lafage, Lukasiniho, Tasman Dixon, Mathew Wedel, Mali’o Kodis, image from
the “Proceedings of the Zoological Society of London”, Gareth Monger,
Collin Gross, Mali’o Kodis, image from the Smithsonian Institution,
Diego Fontaneto, Elisabeth A. Herniou, Chiara Boschetti, Manuela
Caprioli, Giulio Melone, Claudia Ricci, and Timothy G. Barraclough
(vectorized by T. Michael Keesey), Timothy Knepp of the U.S. Fish and
Wildlife Service (illustration) and Timothy J. Bartley (silhouette),
Mark Hofstetter (vectorized by T. Michael Keesey), Rebecca Groom,
Smokeybjb, Jagged Fang Designs, Beth Reinke, T. Michael Keesey (after
James & al.), T. Tischler, Leon P. A. M. Claessens, Patrick M. O’Connor,
David M. Unwin, T. Michael Keesey (photo by Sean Mack), Sarah Werning,
Iain Reid, Rebecca Groom (Based on Photo by Andreas Trepte),
Benjamint444, Original scheme by ‘Haplochromis’, vectorized by Roberto
Díaz Sibaja, Frank Förster, Kai R. Caspar, Pete Buchholz, Steven Coombs
(vectorized by T. Michael Keesey), Maija Karala, FunkMonk, Original
drawing by Dmitry Bogdanov, vectorized by Roberto Díaz Sibaja, Kimberly
Haddrell, Qiang Ou, Scott Reid, Alex Slavenko, Maxime Dahirel, Michelle
Site, Sharon Wegner-Larsen, Milton Tan, Matt Martyniuk (modified by
Serenchia), Obsidian Soul (vectorized by T. Michael Keesey), nicubunu,
NOAA Great Lakes Environmental Research Laboratory (illustration) and
Timothy J. Bartley (silhouette), Ferran Sayol, C. Camilo
Julián-Caballero, Nobu Tamura (vectorized by T. Michael Keesey), Nobu
Tamura, modified by Andrew A. Farke, Scarlet23 (vectorized by T. Michael
Keesey), SauropodomorphMonarch, Henry Lydecker, Noah Schlottman, photo
from Casey Dunn, Mariana Ruiz Villarreal, Mali’o Kodis, photograph by
Melissa Frey, Terpsichores, NASA, Robbie N. Cada (vectorized by T.
Michael Keesey), Alexander Schmidt-Lebuhn, Kamil S. Jaron, Dmitry
Bogdanov (vectorized by T. Michael Keesey), Darren Naish, Nemo, and T.
Michael Keesey, Nina Skinner, Steven Coombs, Nicolas Huet le Jeune and
Jean-Gabriel Prêtre (vectorized by T. Michael Keesey), Elizabeth Parker,
Davidson Sodré, Jaime Headden, Neil Kelley, Dein Freund der Baum
(vectorized by T. Michael Keesey), M Kolmann, Mali’o Kodis, image from
the Biodiversity Heritage Library, Becky Barnes, V. Deepak, Josefine
Bohr Brask, Martin R. Smith, Darren Naish (vectorize by T. Michael
Keesey), Robbie N. Cada (modified by T. Michael Keesey), Lauren
Anderson, Meliponicultor Itaymbere, Anthony Caravaggi, Benchill, Cesar
Julian, Gopal Murali, Roberto Díaz Sibaja, \[unknown\], Mathilde
Cordellier, xgirouxb, Unknown (photo), John E. McCormack, Michael G.
Harvey, Brant C. Faircloth, Nicholas G. Crawford, Travis C. Glenn, Robb
T. Brumfield & T. Michael Keesey, A. R. McCulloch (vectorized by T.
Michael Keesey), Xavier Giroux-Bougard, Harold N Eyster, Duane Raver
(vectorized by T. Michael Keesey), Dean Schnabel, Mathew Stewart,
Shyamal, Carlos Cano-Barbacil, Konsta Happonen, Ingo Braasch,
SecretJellyMan - from Mason McNair, Fcb981 (vectorized by T. Michael
Keesey), Claus Rebler, Christine Axon, Original drawing by Antonov,
vectorized by Roberto Díaz Sibaja, Michele Tobias, Amanda Katzer, Renato
de Carvalho Ferreira, Jessica Anne Miller, David Tana, Katie S. Collins,
Noah Schlottman, photo by David J Patterson, Mike Hanson, Taro Maeda,
Jaime A. Headden (vectorized by T. Michael Keesey), Pearson Scott
Foresman (vectorized by T. Michael Keesey), David Orr, Andrew A. Farke,
Nancy Wyman (photo), John E. McCormack, Michael G. Harvey, Brant C.
Faircloth, Nicholas G. Crawford, Travis C. Glenn, Robb T. Brumfield & T.
Michael Keesey, Robert Gay, modified from FunkMonk (Michael B.H.) and T.
Michael Keesey., Chris Hay, Jaime Headden, modified by T. Michael
Keesey, Ellen Edmonson and Hugh Chrisp (vectorized by T. Michael
Keesey), Mattia Menchetti / Yan Wong, Aviceda (photo) & T. Michael
Keesey, Oscar Sanisidro, CNZdenek, Robert Bruce Horsfall, vectorized by
Zimices, Jack Mayer Wood, Mathieu Basille, Michael B. H. (vectorized by
T. Michael Keesey), Henry Fairfield Osborn, vectorized by Zimices, I.
Sácek, Sr. (vectorized by T. Michael Keesey), FJDegrange, JCGiron,
Dmitry Bogdanov, Stanton F. Fink (vectorized by T. Michael Keesey),
Smokeybjb (vectorized by T. Michael Keesey), Ghedoghedo (vectorized by
T. Michael Keesey), Emily Willoughby, T. Michael Keesey (from a
photograph by Frank Glaw, Jörn Köhler, Ted M. Townsend & Miguel Vences),
Matt Martyniuk (modified by T. Michael Keesey), Ludwik Gasiorowski,
Michael Day, Stacy Spensley (Modified), Zachary Quigley, S.Martini,
Manabu Sakamoto, Andreas Hejnol, Todd Marshall, vectorized by Zimices,
Noah Schlottman, Rene Martin, Lisa Byrne, Chris Jennings (Risiatto),
Nobu Tamura, Tyler Greenfield and Scott Hartman, Juan Carlos Jerí,
FunkMonk \[Michael B.H.\] (modified by T. Michael Keesey), Noah
Schlottman, photo by Casey Dunn, Felix Vaux, Mali’o Kodis, photograph by
P. Funch and R.M. Kristensen

## Detailed credit:

|     | Image X Coord | Image Y Coord | Credit                                                                                                                                                                |
| --: | ------------: | ------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
|   1 |    296.620016 |    329.315759 | T. Michael Keesey                                                                                                                                                     |
|   2 |    349.948808 |    585.599713 | Arthur S. Brum                                                                                                                                                        |
|   3 |    410.210785 |    452.099999 | Julio Garza                                                                                                                                                           |
|   4 |    201.280434 |    741.783088 | Birgit Lang                                                                                                                                                           |
|   5 |    817.939117 |    240.978766 | Elisabeth Östman                                                                                                                                                      |
|   6 |    518.887883 |    553.847089 | Mo Hassan                                                                                                                                                             |
|   7 |    578.728556 |    219.074088 | NA                                                                                                                                                                    |
|   8 |    514.436316 |    136.832590 | Scott Hartman                                                                                                                                                         |
|   9 |    511.656512 |    381.912336 | NA                                                                                                                                                                    |
|  10 |    905.945468 |    179.650106 | Zimices                                                                                                                                                               |
|  11 |    907.872296 |    587.218699 | Matt Crook                                                                                                                                                            |
|  12 |    531.910604 |    692.118648 | Steven Traver                                                                                                                                                         |
|  13 |     76.913775 |    326.162606 | Margot Michaud                                                                                                                                                        |
|  14 |    106.418438 |     76.160976 | Matthew E. Clapham                                                                                                                                                    |
|  15 |    462.122913 |     47.046577 | Zimices                                                                                                                                                               |
|  16 |    765.602366 |    736.521594 | Sergio A. Muñoz-Gómez                                                                                                                                                 |
|  17 |    779.095084 |     68.810937 | T. Michael Keesey                                                                                                                                                     |
|  18 |    695.933702 |    402.460929 | Nobu Tamura, vectorized by Zimices                                                                                                                                    |
|  19 |    941.604256 |    344.663705 | Didier Descouens (vectorized by T. Michael Keesey)                                                                                                                    |
|  20 |    733.310378 |    331.387537 | Chase Brownstein                                                                                                                                                      |
|  21 |    291.846255 |    656.310977 | Tracy A. Heath                                                                                                                                                        |
|  22 |    912.504971 |    453.818378 | Christoph Schomburg                                                                                                                                                   |
|  23 |    657.717243 |    146.323652 | Mali’o Kodis, drawing by Manvir Singh                                                                                                                                 |
|  24 |    114.350411 |    512.169926 | Margot Michaud                                                                                                                                                        |
|  25 |    715.732959 |    550.489286 | Walter Vladimir                                                                                                                                                       |
|  26 |    794.771422 |    110.231414 | Ray Simpson (vectorized by T. Michael Keesey)                                                                                                                         |
|  27 |    329.588247 |     69.530986 | Steven Traver                                                                                                                                                         |
|  28 |    197.999481 |    592.413800 | Margot Michaud                                                                                                                                                        |
|  29 |    897.182469 |     91.350759 | Zimices                                                                                                                                                               |
|  30 |    616.838951 |     66.809447 | Chris huh                                                                                                                                                             |
|  31 |    831.077913 |    681.050470 | Steven Traver                                                                                                                                                         |
|  32 |    740.490569 |    502.096310 | Margot Michaud                                                                                                                                                        |
|  33 |    809.406082 |    403.764490 | Matt Crook                                                                                                                                                            |
|  34 |    343.073557 |    542.317193 | Gabriela Palomo-Munoz                                                                                                                                                 |
|  35 |    665.974886 |    726.335665 | mystica                                                                                                                                                               |
|  36 |     66.798152 |    721.101080 | NA                                                                                                                                                                    |
|  37 |    585.178220 |    467.913865 | Lafage                                                                                                                                                                |
|  38 |    113.631795 |    213.365841 | Steven Traver                                                                                                                                                         |
|  39 |    506.809943 |    320.692972 | Lukasiniho                                                                                                                                                            |
|  40 |     65.876437 |    157.396497 | Tasman Dixon                                                                                                                                                          |
|  41 |    772.627827 |    611.795543 | Matt Crook                                                                                                                                                            |
|  42 |    382.526154 |    756.556860 | Mathew Wedel                                                                                                                                                          |
|  43 |    976.215407 |    647.627555 | Mali’o Kodis, image from the “Proceedings of the Zoological Society of London”                                                                                        |
|  44 |    670.568956 |    245.210389 | Gareth Monger                                                                                                                                                         |
|  45 |    223.731451 |    106.994730 | Gareth Monger                                                                                                                                                         |
|  46 |    161.822134 |    299.010589 | Collin Gross                                                                                                                                                          |
|  47 |    618.694754 |    585.739331 | Mali’o Kodis, image from the Smithsonian Institution                                                                                                                  |
|  48 |    157.727723 |    403.093658 | Diego Fontaneto, Elisabeth A. Herniou, Chiara Boschetti, Manuela Caprioli, Giulio Melone, Claudia Ricci, and Timothy G. Barraclough (vectorized by T. Michael Keesey) |
|  49 |    101.173235 |    677.605708 | Timothy Knepp of the U.S. Fish and Wildlife Service (illustration) and Timothy J. Bartley (silhouette)                                                                |
|  50 |    708.173514 |     61.461724 | Mark Hofstetter (vectorized by T. Michael Keesey)                                                                                                                     |
|  51 |    428.510655 |    613.934680 | Rebecca Groom                                                                                                                                                         |
|  52 |    908.597358 |    751.192848 | Smokeybjb                                                                                                                                                             |
|  53 |    888.378609 |    224.499049 | Jagged Fang Designs                                                                                                                                                   |
|  54 |    596.746180 |    405.511944 | Beth Reinke                                                                                                                                                           |
|  55 |     26.479078 |    427.559269 | NA                                                                                                                                                                    |
|  56 |    494.572407 |    230.840783 | Mathew Wedel                                                                                                                                                          |
|  57 |    448.296853 |    585.607116 | Scott Hartman                                                                                                                                                         |
|  58 |    941.078490 |    266.845438 | T. Michael Keesey (after James & al.)                                                                                                                                 |
|  59 |     83.614524 |    276.193816 | T. Tischler                                                                                                                                                           |
|  60 |    411.078685 |    368.542170 | Leon P. A. M. Claessens, Patrick M. O’Connor, David M. Unwin                                                                                                          |
|  61 |    389.623208 |    732.195045 | Jagged Fang Designs                                                                                                                                                   |
|  62 |    391.218045 |    665.815946 | T. Michael Keesey (photo by Sean Mack)                                                                                                                                |
|  63 |    984.952733 |    421.539220 | Sarah Werning                                                                                                                                                         |
|  64 |    849.876214 |    519.245258 | Margot Michaud                                                                                                                                                        |
|  65 |    963.802957 |     97.052774 | Iain Reid                                                                                                                                                             |
|  66 |    300.559823 |    116.124745 | Rebecca Groom (Based on Photo by Andreas Trepte)                                                                                                                      |
|  67 |    949.788725 |    621.312661 | Benjamint444                                                                                                                                                          |
|  68 |    181.109669 |    342.784997 | Original scheme by ‘Haplochromis’, vectorized by Roberto Díaz Sibaja                                                                                                  |
|  69 |    111.452917 |    373.246896 | Gareth Monger                                                                                                                                                         |
|  70 |    229.570653 |     37.421453 | Sarah Werning                                                                                                                                                         |
|  71 |    988.741610 |    474.404405 | Steven Traver                                                                                                                                                         |

NEWS.md
# impac (development version)

# impac 0.1.0

* Added ability to put custom metadata in the metadata output (by returning a list from an image 
generator, rather than just an image)
* Github actions regenerates Readme daily including two packed images (new images daily!)
* Added a `NEWS.md` file to track changes to the package

# impac 0.0.0.9000

* Main function `impac` fully functional


R/impac.R
#' Make a packed image mosaic
#'
#' A simple greedy algorithm tries to pack as many images
#' into a larger image as possible, taking into account
#' transparency, if available (recommended).
#'
#' @param im Can be either a character vector of image
#' file names (format must be compatible with [imager::load.image()]),
#' a list of [`imager::cimg`] objects, or a function that generates
#' an image when evaluated. The function can take a single argument,
#' which is the current iteration of the packing algorithm. Can also
#' be specified as an `rlang` style lambda syntax (see [rlang::as_function()]).
#' @param width Width in pixels of produced image
#' @param height Height in pixels of produced image
#' @param mask An optional masking image.
#' @param weights Vector of Weights to apply to each image. Higher weighted
#' images will  be packed first and so will tend to be larger. This vector
#' will be recycled.
#' @param preferred An alternate way to specify images to pack first, as
#' a character vector of names or file names (only works if `im` is a
#' vector of image file name or a list of [`imager::cimg`] objects).
#' @param max_num_tries Maximum number of times to try packing an image
#' onto the canvas before giving up.
#' @param scales A vector of starting scaling factors to randomly choose
#' from for each image.
#' @param scale_fun An function that takes three arguments,
#' which correspond to the current vector of scaling factors, the current
#' iteration of the algorithm, and the count of the number of packed images
#' so far, respectively (e.g. `f(s, i, c)`), and returns a new vector of
#' scaling factors to use.
#' @param max_images The maximum number of images to pack before stopping.
#' @param min_scale The minimum scale factor to use. If the algorithm
#' generates a scale factor this small (via `scale_fun`), packing will stop.
#' @param bg The background colour for the campus, default: "transparent"
#' @param show_every Show the intermediate packed image after every
#' `show_every` images are packed. Set to 0 to not show intermediates.
#' @param progress Should progress be printed as the algorithm runs?
#' @param start_image An optional image to start the packing with. If not
#' `NULL`, the `width` and `height` arguments will be ignored and the
#' dimensions of the starting image used instead. Can be an `imager::cimg`
#' object, a path to an image in png or jpg format or an `impac` object.
#' @param ... Further arguments passed on the `im`, if it is function.
#'
#' @return A packed image mosaic, as a [`imager::cimg`] object.
#' @export
#'
#' @importFrom imager %inr%
#' @importFrom grDevices col2rgb
#' @importFrom stats runif
#'
#' @examples
#' plot(
#'   impac(
#'     function(i) imager::draw_circle(
#'       imager::imfill(500, 500, val = c(0, 0, 0, 0)),
#'       250, 250, radius = runif(1, 150, 250),
#'       color = matrix(grDevices::col2rgb(sample(grDevices::rainbow(100), 1), alpha = TRUE), nrow = 1)
#'     ),
#'     width = 400, height = 400,
#'     max_images = 10, bg = "white"
#'   )$image
#' )
impac <- function(im, width = 1024, height = 800,
                  mask = NULL,
                  weights = NULL,
                  preferred = NULL,
                  max_num_tries = 100,
                  scales = c(rep(0.5, 2), rep(0.25, 4), rep(0.15, 8)),
                  scale_fun = function(s, i, c) {
                    if(c < (i * 0.5)) {
                      mscale <- min(s)
                      c(s, rep(mscale / 2, floor(1 / mscale)))
                    } else {
                      scales
                    }
                  },
                  max_images = 1000,
                  min_scale = 0.05,
                  bg = "transparent",
                  show_every = 25,
                  progress = TRUE,
                  start_image = NULL,
                  ...) {


  settings <- lapply(as.list(match.call()), function(x) evalq(x))
  settings["start_image"] <- NULL

  impac_env$latest_args <- settings[-1]

  bg_col <- as.vector(col2rgb(bg)) / 255
  if(!is.null(start_image)) {
    if(inherits(start_image, "character")) {
      canvas <- convert_to_rgba(imager::load.image(start_image))
    }
    if(inherits(start_image, "cimg")) {
      canvas <- convert_to_rgba(start_image)
    }
    if(inherits(start_image, "impac")) {
      canvas <- convert_to_rgba(start_image$image)
    }
    width <- imager::width(canvas)
    height <- imager::height(canvas)
  } else {
    canvas <- imager::imfill(x = width, y = height,
                             val = c(0, 0, 0, 0))
  }

  if(!is.null(mask)) {

    if(inherits(mask, "character")) {
      mask <- imager::load.image(mask)
    }

    canvas <- imager::resize(canvas,
                             width(mask),
                             height(mask))
    mask <- mask %>%
      imager::grayscale() %>%
      imager::threshold(0.5)

  } else {

    mask <- imager::channel(canvas, 4)

  }

  if(rlang::is_formula(im)) {
    im <- rlang::as_function(im)
  }

  if(!rlang::is_function(im)) {
    if(rlang::is_list(im)) {
      if(!all(sapply(im, class) == "cimg")) {
        stop("If im is a list, it must be a list of cimg objects (see imager documentation for details).")
      }
      im_type <- "cimgs"
    } else {
      if(!rlang::is_character(im)) {
        stop("If im is a vector it must be a character vector of file names.")
      }
      im_type <- "filenames"
    }
    num_images <- length(im)
  } else {
    num_images <- max_images
    im_type <- "function"
  }

  if(!is.null(preferred) & is.null(weights)) {
    if(!rlang::is_vector(preferred) & !(rlang::is_integer(preferred) | rlang::is_integer(preferred))) {
      rlang::abort("If preferred is specified it must be a character vector of filenames or an integer vector of indices")
    }
    weights <- rep(0.001, num_images)
    if(im_type == "filenames" & rlang::is_character(preferred)) {
      names(weights) <- im
    }
    weights[preferred] <- 1
    num_preferred <- length(preferred)
  } else {
    num_preferred <- floor(0.2 * num_images)
  }

  if(is.null(weights)) {
    weights <- rep(1, num_images)
  }

  if(im_type != "function") {
    im <- sample(im, prob = weights)
  }

  if(!is.null(start_image) & inherits(start_image, "impac")) {
    image_map <- start_image$meta
  } else {
    image_map <- data.frame(NULL)
  }
  count <- 0

  if(progress) {
    total <- ifelse(im_type == "function", "?", as.character(num_images))
    format <- paste0(":spin (:current/", total, " images packed. Packing at :tick_rate images per second. Time elapsed: :elapsedfull")
    pr <- progress::progress_bar$new(format = format, total = NA)
  }

  for(i in seq_len(num_images)) {

    if(im_type == "filenames") {
      img <- imager::load.image(im[[i]])
      img <- convert_to_rgba(img)
    }
    if(im_type == "cimgs") {
      img <- im[[i]]
    }
    if(im_type == "function") {
      img <- im(i, ...)
      if(rlang::is_list(img)) {
        meta <- img[-1]
        img <- img[[1]]
      } else {
        meta <- NULL
      }
      #img <- im(i)
    }

    success <- FALSE
    for(j in seq_len(max_num_tries)) {

      x <- runif(1, 0, width)
      y <- runif(1, 0, height)
      scale <- sample(scales, 1)

      if(i <= num_preferred) {
        scale <- min(1, scale * 2)
      }

      w <- as.integer(imager::width(img) * scale)
      h <- as.integer(imager::height(img) * scale)

      if(w < 3 | h < 3) {
        next
      }

      if((x < (w * 0.5)) | (x > (imager::width(canvas) - w * 0.5))) {
        next
      }
      if((y < (h * 0.5)) | (y > (imager::height(canvas) - h * 0.5))) {
        next
      }

      resized_img = imager::resize(img, w, h, interpolation_type = 6)
      resized_img <- imager::imchange(resized_img, ~ . < 0, ~ 0)
      resized_img <- imager::imchange(resized_img, ~ . > 1, ~ 1)

      w <- imager::width(resized_img)
      h <- imager::height(resized_img)

      xmin <- as.integer(max(1, x - 0.5 * w))
      ymin <- as.integer(max(1, y - 0.5 * h))
      xr <- c(xmin, xmin + w - 1L)
      yr <- c(ymin, ymin + h - 1L)

      sub_mask <- mask[xr[1]:xr[2], yr[1]:yr[2], , , drop = FALSE]
      img_mask <- imager::channel(resized_img, 4)

      composite <- any(imager::parmin(list(sub_mask, img_mask)) > 0)

      if(composite) {
        if(progress) {
          pr$tick(0)
        }
        next
      }

      ## paste image into canvas
      new_img <- imager::add(list(canvas[xr[1]:xr[2], yr[1]:yr[2], , , drop = FALSE], resized_img))
      canvas[xr[1]:xr[2], yr[1]:yr[2], , ] <- new_img
      ## regenerate mask
      mask <- imager::channel(canvas, 4)
      success <- TRUE
      dat <- do.call(data.frame, c(list(x = x, y = y, scale = scale, image = i), meta))
      image_map <- rbind(image_map, dat)

      break

    }

    if(success) {
      count <- count + 1
      if(progress) {
        pr$tick()
      }
      if(show_every != 0) {
        if(count %% show_every == 0) {
          if(bg != "transparent") {
            plot(imager::flatten.alpha(canvas, bg = bg))
          } else {
            plot(canvas)
          }
        }
      }
      impac_env$saved_image <- canvas
      impac_env$meta <- image_map
    } else {
      scales <- scale_fun(scales, i, count)
    }

    mscale = min(scales)
    if(mscale < min_scale) {
      message("Packing stopped since not enough empty space is left.")
      break;
    }

  }

  if(bg != "transparent") {
    canvas <- imager::flatten.alpha(canvas, bg = bg)
  }

  if(progress) {
    pr$terminate()
  }

  res <- list(image = canvas, meta = image_map, args = settings[-1])
  attr(res, "env") <- parent.frame()
  class(res) <- "impac"

  return(res)

}
